pkg/controllermanager/controller/controllerinstallation/controllerinstallation_control.go

```diff
@@ -0,0 +1,398 @@
+		} else {
+			conditionInstalled = helper.UpdatedCondition(conditionInstalled, corev1.ConditionFalse, "DeletionFailed", fmt.Sprintf("Deletion of old resources failed: %+v", err))
+		}
+		return err
+	}
+	conditionInstalled = helper.UpdatedCondition(conditionInstalled, corev1.ConditionFalse, "DeletionSuccessful", "Deletion of old resources succeeded.")
+
+	_, err = kutil.TryUpdateControllerInstallationWithEqualFunc(c.k8sGardenClient.GardenCore(), retry.DefaultBackoff, controllerInstallation.ObjectMeta, func(c *corev1alpha1.ControllerInstallation) (*corev1alpha1.ControllerInstallation, error) {
+		finalizers := sets.NewString(c.Finalizers...)
+		finalizers.Delete(FinalizerName)
+		c.Finalizers = finalizers.UnsortedList()
+		return c, nil
+	}, func(cur, updated *corev1alpha1.ControllerInstallation) bool {
+		return !sets.NewString(cur.Finalizers...).Has(FinalizerName)
+	})
+	return err
+}
+
+func (c *defaultControllerInstallationControl) updateConditions(controllerInstallation *corev1alpha1.ControllerInstallation, conditions ...corev1alpha1.Condition) (*corev1alpha1.ControllerInstallation, error) {
+	return kutil.TryUpdateControllerInstallationStatusWithEqualFunc(c.k8sGardenClient.GardenCore(), retry.DefaultBackoff, controllerInstallation.ObjectMeta,
+		func(controllerInstallation *corev1alpha1.ControllerInstallation) (*corev1alpha1.ControllerInstallation, error) {
+			controllerInstallation.Status.Conditions = conditions
+			return controllerInstallation, nil
+		}, func(cur, updated *corev1alpha1.ControllerInstallation) bool {
+			return equality.Semantic.DeepEqual(cur.Status.Conditions, updated.Status.Conditions)
+		},
+	)
+}
+
+func (c *defaultControllerInstallationControl) isResponsible(controllerInstallation *corev1alpha1.ControllerInstallation) (bool, error) {
+	controllerRegistration, err := c.controllerRegistrationLister.Get(controllerInstallation.Spec.RegistrationRef.Name)
+	if err != nil {
+		return false, err
+	}
+
+	return controllerRegistration.Spec.Deployment.Type == "helm", nil
+}
+
+func (c *defaultControllerInstallationControl) cleanOldResources(k8sSeedClient kubernetes.Interface, controllerInstallation *corev1alpha1.ControllerInstallation, newResourcesSet sets.String) (bool, error) {
+	providerStatus := controllerInstallation.Status.ProviderStatus
+	if providerStatus == nil {
+		return false, nil
+	}
+
+	var oldResources DeployedResources
+	if err := json.Unmarshal(providerStatus.Raw, &oldResources); err != nil {
+		return false, err
+	}
+
+	var (
+		deleted = true
+		result  error
+	)
+
+	for _, oldResource := range oldResources.Resources {
+		if !newResourcesSet.Has(objectReferenceToString(oldResource)) {
+			obj := &unstructured.Unstructured{}
+			obj.SetAPIVersion(oldResource.APIVersion)
+			obj.SetKind(oldResource.Kind)
+			obj.SetNamespace(oldResource.Namespace)
+			obj.SetName(oldResource.Name)
+
+			if err := k8sSeedClient.Client().Delete(context.TODO(), obj); err != nil {
```

Comment:
It seems you're duplicating some code here - this is mostly the same as the `Apply` code, it's just `Delete`ing instead of creating or updating. Can't we have an `Unapply` method in the `Applier`? On the way while doing so, you might even have to factor out the reading of the objects from a manifest just like @zanetworker did it in #540.

Reply:
Added a TODO for later usage once #624 is merged.